### PR TITLE
[RDY] Tiling and Workspace fixes

### PR DESCRIPTION
**Pull request by ramirezmike**
I'll try to keep this original post updated as I get feedback/make changes

This pull request contains the following commits...

1. 05162d2 Fix: workspaces remain ordered by ID when moved across monitor
    I noticed that moving workspaces back and forth across monitors changes the order of them. If you have the following...
      monitor1            monitor2
      1 2 3  7                 5 6
and then moving workspace  5 back to monitor 1 it ends up like this
      monitor1            monitor2
      1 2 3  7 5                6     
This commit keeps the order so it ends up "1 2 3 5 7"

2. 31cc8d2 Feat: made right-alt usage configurable
This is in regards to #210 and just adds a config value to allow usage of the right alt. This is just an initial attempt at it, open to suggestions on a better implementation. Still need to add documentation, but that's dependent on if we decide this is a good idea or not. I'm also not sure exactly where in the doc-gen project that the documentation for config values goes?

3. 30aeb96 Fix: prevent unmanaging windows from inactive workspaces
  This tripped me up a bit a couple times but basically if you switch to another workspace it's possible depending on what windows are currently visible that you still have focus on a window that was on the workspace you just left... and if you toggle to manage/unmanage a window at this point it unmanages the window from the workspace that isn't currently visible. This is more likely to happen when working across multiple monitors and losing track of what window on what workspace you think is focused. 

    This commit just adds some logic to prevent unmanaging the window if it's not on the focused workspace since it's unlikely the user wants that to happen. I also went ahead and renamed a couple functions like find_window and find_grid as their names were incorrect; find_window returns a grid and find_grid returns a display, so I renamed them to find_grid_containing_window and find_grid_display. 

4. a2f31d5 Fix: #211 columns and rows merge
This should be pretty straight forward and is just adding logic for a scenario I documented in #211 that I came across. Also added unit tests for this one.

5.  5d5c617 Feat: Added fullscreen indicator component
This is a quality-of-life improvement for me personally. I often jump back and forth between workspaces and get confused about how I left them. This is to set up a string indicator component that displays in the bar when the current workspace has a window fullscreened. That way I can quickly tell if I have other windows hiding behind the fullscreened one. I added documentation for this one.

6.  1add07e Fix: Re-add missing ignore_fullscreen_actions checks
This is pretty straight forward, there were a couple actions that had their ignore_fullscreen_actions checks missed during our recent config changes. This just re-adds them to make that config value work a bit better.

7. 65c0159 Fix: #222 access denied saving grid
This was just fixing me being dumb trying to save the grid to a folder. When in debug mode it was trying to save to ./log rather than an actual file so I updated it to save to ./log/workspace.grid when running in debug mode. Release mode should be unaffected.

8. 86b7f83 Feat: added workspace_to_workspace nogscript binding
This adds a nogscript workspace function to move an entire workspace to another workspace as long as the target workspace is empty. I find this useful as it's easier for me to hit my main workspaces 1-4 while leaving the rest of the workspaces as "backburner" workspaces. I kept it at only working if the target is empty to avoid any weirdness caused by combining workspaces; I figure that'd be something much more complex and not user-friendly.

9. b076b3f Feat: updated deadlock logging to log all locked threads' backtraces
While working on all this I kept running into deadlocks and noticed that the deadlock detection just spits out the backtrace of one of the stacks that is locked. I updated this to show all threads that are deadlocked. Which helped me with...

**UPDATED BELOW Feb 03**

10.  d5f286a Fix: Prevent deadlocks when updating config values
While working on this pullrequest I kept running into deadlocks with the keybindings and I isolated them to two scenarios. This commit is just to prevent the deadlocks caused by the first scenario.

Both happen during hot-reloading. To re-create, try toggling the app_bar & taskbar repeatedly at random intervals.

Deadlock A (fixed in this commit):
hot_reload::update_config locks **state** and then calls KbManager.get_mode which attempts to lock "**inner**"

bottom loop of KbManager.start locks **inner** then tries to lock **state**

These two lock each other. I resolved this one by dropping the "inner" keyboard object in the bottom loop of KbManager once it's no longer needed.

Deadlock B (not fixed):
The other one is a little bit weirder but involves the [Windows SendMessageA function](https://docs.microsoft.com/en-us/windows/win32/api/winuser/nf-winuser-sendmessagea). This happens when trying to close the app_bar; it calls SendMessageA on the app_bar window which "does not return until the window procedure has processed the message." Which, wouldn't be a big deal but...

What I've seen happen is that closing the app_bar locks the **state**, then it trigger this SendMessageA which fires off a window close event. Meanwhile, a separate thread is attempting to draw the app_bar and is attempting to lock the **state** repeatedly (WindowEvent::Draw in twm/src/bar/create.rs). So, it deadlocks with the **state** already locked in the SendMessageA call and the handler waiting on the Draw event to finish before handling the SendMessageA's close event and the Draw event waiting on the **state** to be unlocked. I was able to prevent this deadlock by using [PostMessageA](https://docs.microsoft.com/en-us/windows/win32/api/winuser/nf-winuser-postmessagea) instead which queues the request rather than waiting on it.

Unfortunately, while using it throughout the day I noticed this introduced flashing artifact of the components in the app_bar and occasionally the app_bars would close. I'm not 100% sure why but once I reverted this portion of the commit those bugs disappeared. 

11. 44362f3 Feat: Updated documentation
Just updated the one doc line for workspace_to_workspace and removed a debug I left in.

12. cd413f5 Formatting
just running format on everything.
 

- [X] Add a move workspace to another workspace, i.e, moving all the tiles on workspace 3 to workspace 9
- [X] Re-add missing ignore_fullscreen_actions checks
- [X] Do something about #222

### CODE REVIEW

File: twm/src/keybindings.rs [hunk 1]
```diff
@@ -140,6 +142,9 @@ impl KbManager {
             .send(ChanMessage::ChangeMode(mode.clone()))
             .expect("Failed to change mode of kb manager");
     }
+    pub fn update_configuration(&self, config: &Config) {
```
Comment (ramirezmike) [inline twm/src/keybindings.rs:145]: this is probably the most debatable part of this commit. I needed a way to flip the allow_right_alt value that is used in the loop further below, so I added this function that would be used for any other hot_reload type configuration changes in the future.

File: doc-gen/nogscript/nog/workspace.ns [hunk 2]
```diff
@@ -6,6 +6,11 @@ extern fn change(id)
 /// @param id Number
 extern fn move_to_monitor(id)
 
+/// Moves the focused workspace to the workspace that has the given index.
```
Comment (TimUntersberger) [inline doc-gen/nogscript/nog/workspace.ns:9]: Should probably explain what "moving" a workspace to another workspace means.

**Review by TimUntersberger — CHANGES_REQUESTED** [review]

File: doc-gen/nogscript/nog/workspace.ns (hunk 2 above)
Comment (ramirezmike) [reply doc-gen/nogscript/nog/workspace.ns]: Updated this:

/// Moves all windows in the focused workspace to the workspace that has 
/// the given index while maintaining relative sizes and positions. 
/// The target workspace must be empty. This allows you to transfer
/// an entire workspace to an empty workspace of a different number.
/// @param id Number

**Comment by ramirezmike** [comment]
Going to change this from draft and mark it as "RDY." Still open to suggestions but I think this is about as much as I'll add to this pull request.

**Comment by TimUntersberger** [comment]
@ramirezmike the PR is still a draft.

**Pull request merged**